Make it possible to disallow slow queries (or warn user about the problem)

msm-code opened the pull request:
<!-- Thank you for contributing! -->
<!-- Please fill this template before submitting your PR (fill the boxes using "x") -->

**Your checklist for this pull request**
- [x] I've read the [contributing guideline](https://github.com/CERT-Polska/mquery/blob/master/CONTRIBUTING.md).
- [x] I've tested my changes by building and running mquery, and testing changed functionality (if applicable)
- [ ] I've added automated tests for my change (if applicable, optional)
- [ ] I've updated documentation to reflect my change (if applicable) - not yet, I'll add a separate issue for documentation update

**What is the current behaviour?**

#308 

>Currently, people who don't understand how mquery and mwdb work behind the scene can execute very bad yara queries and it'll be (a) expensive (b) inefficient (c) take forever to run.

**What is the new behaviour?**

![image](https://user-images.githubusercontent.com/7026881/209887840-2305e3c1-878a-439f-8de7-31c2310d7a40.png)

If configuration option `query_disallow_degenerate` is set to "true", users won't be able to run any queries that would end up doing a whole backend scan. Of course they can still run very inefficient queries like `"aaa" | "bbb" | ccc" | "ddd"` or just `{00 00 00}`, so this won't help in every case, but that's all we can do at this stage (others will be handled by #297).

I know this is not exactly what #308 asked for. I wonder if anyone really wants to use the "soft" mode - warn only, don't reject the query. If not, I think it's good as just a configuration option.

**Closing issues**

<!-- Add in issue numbers related to this PR, if applicable -->

fixes #308

ITAYC0HEN commented on src/app.py:
```diff
@@ -335,11 +335,28 @@ def query(
+                "Some of the rules would require a Yara scan of every indexed "
+                "file, and this is not allowed by this instance. "
+                f"Problematic rules: {degenerate_rule_names}. "
+                f"Read {doc_url} for more details."
```
link is not clickable, I guess it should :)

msm-code replied:
Not possible with an error message (without severe hacks, like putting HTML code into a Python exception, or - even worse - parsing the error message on the frontend.

I guess I'll do it the harder way, and put the related alert logic into the frontend.

ITAYC0HEN commented:
> I wonder if anyone really wants to use the "soft" mode - warn only

Sometimes users really want to just list all the files that has a certain [imphash](https://yara.readthedocs.io/en/latest/modules/pe.html#c.imphash) value, and it's very important for their research. (i hope this falls under "degenerate query")

```
pe.imphash() == "b8bb385806b89680e13fc0cf24f4431e"
``` 


On other times, users don't really understand the problem.


Combining warn-only mode, **and** #297 will make sure that even if the user really wants to proceed, they can and it will not be too bad for the system (thanks to the file-limit)

nazywam commented on src/app.py:
```diff
@@ -335,11 +335,28 @@ def query(
                 rule_author=rule.author,
                 is_global=rule.is_global,
                 is_private=rule.is_private,
+                is_degenerate=rule.parse().is_degenerate,
```
We could probably call `rule.parse()` just once, somewhere above? In case it gets expensive in the future

msm-code replied:
Performance hit is negliglible, and even if it wasn't, it's already cached:

```python
    def parse(self) -> UrsaExpression:
        if self.__parsed is None:
            self.__parsed = self.__parse_internal()
        return self.__parsed
```

I worry more about readability. There's no way to assign it to a temporary variable in a list comprehension, but I can rewrite this all to an explicit for loop if you prefer :thinking: It would look like this:

```python
response_rules = []
for rule in rules:
    parsed_rule = rule.parse()
    response_rules.append(
        ParseResponseSchema(
            rule_name=rule.name,
            rule_author=rule.author,
            is_global=rule.is_global,
            is_private=rule.is_private,
            is_degenerate=parsed_rule.is_degenerate,
            parsed=parsed_rule.query,
        )
    )
return response_rules
```

I prefer the current version, but I'm not overly attached to it if you prefer this one.

nazywam commented on src/app.py:
```diff
@@ -335,11 +335,28 @@ def query(
 
+    degenerate_rules = [r.name for r in rules if r.parse().is_degenerate]
+    disallow_degenerate = db.get_mquery_config_key("query_disallow_degenerate")
+    if degenerate_rules and (disallow_degenerate == "true"):
```
Seems like `disallow_degenerate == "true"` should be somewhere lower in the abstraction layer :thinking: (just a thought)

nazywam replied:
Also, I don't think the parentheses are needed?
```suggestion
    if degenerate_rules and disallow_degenerate == "true":
```

msm-code replied:
O, and of course you're right about the lower abstraction layer. It's planned and scheduled for the next release: #257.

The current config is a mess anyway, for example I've named it "disallow_degenerate" instead of "allow_degenerate", because empty value is the default, and I want the default to be allowed. Letting storage details leak into config key names is not the best abstraction.

msm-code commented:
> (i hope this falls under "degenerate query")

It does :). No viable ngrams == "degenerate".

> Combining warn-only mode, and https://github.com/CERT-Polska/mquery/issues/297 will make sure that even if the user really wants to proceed, they can and it will not be too bad for the system (thanks to the file-limit)

ACK, thanks for input. I'll rework the code then to just warn the user in the 'non-enforcing' mode.

ITAYC0HEN commented:
So 3 modes? Allow\Ignore, Warn, Prevent\Block?

ITAYC0HEN commented:
it can also be nice if file-limit can be configured to work:
1. always
2. only in warn and block modes
3. only in block mode
4. never

msm-code commented:
> So 3 modes? Allow\Ignore, Warn, Prevent\Block?

I was considering it, but I think just two modes - `Warn` and `Block` are enough. It will be very easy to bypass the warning, and even when we trust users absolutely it may be nice to get an information that rule is inefficient (this can for example happen when copy-pasting rules from the internet - sometimes they look good but have something like `or pe.imphash == '1234'` that ruins it for mquery. I have a WIP docs page about writing good yara rules).

So in short, I don't see the application for `Always` mode, but I'm not fundamentally against it if anyone needs it.

> it can also be nice if file-limit can be configured to work:

Haven't thought of that yet. I don't immediately like the idea, because:

- it adds a dependency between otherwise independent features)
- Degenerate (sorry for the name btw, best I could do) queries are not the only ones hard for mquery. I've tested with https://github.com/Neo23x0/signature-base on a dataset with 750k malicious files, and even though most of the rules returned way less than 1% of dataset to scan, there were some really bad (for mquery) rules. Especially top3 worst rules were "almost as bad" as degenerate ones: https://github.com/Neo23x0/signature-base/blob/master/yara/gen_excel_auto_open_evasion.yar (712806 files), https://github.com/Neo23x0/signature-base/blob/master/yara/crime_ransom_darkside.yar (695732 files) and https://github.com/Neo23x0/signature-base/blob/master/yara/gen_xor_hunting.yar (662770 files).

But this is probably discussion for #297

msm-code commented:
@nazywam @ITAYC0HEN can you re-review?

Changes since the last time:

- I've changed `query_disallow_degenerate` to `query_allow_slow`. Nitpicker in me doesn't like the new name (slow queries are still possible, just harder), but it's probably easier to understand
- If that config key is set to "false" or empty, database will reject the query outright, and link the docs (I've updated them too)
- If that config is set to "true", database will ask the user if they're sure, the query button will change to scary red, and the search will start if the user clicks it again:

![image](https://user-images.githubusercontent.com/7026881/210800254-1a95c997-05bc-46b0-bb4e-2d0b79e80a81.png)

Also I've sneakily fixed broken CSS in the dropdowns.

Things I didn't change:

- Link is still not clickable. After tackling that for a while, I think this would require quite a few changes in the error handling on both server and client (at least if we want to do it correctly). I don't feel it's important enough improvement to justify the number of changes (link is copy&pasteable so the user can easily copy it). Maybe we can make it a (low priority?) issue.
- As per my previous message, there's no "Allow/Ignore" mode, only "Warn" and "Prevent/Block"

yankovs commented on src/app.py:
```diff
@@ -335,11 +335,34 @@ def query(
         ]
 
+    degenerate_rules = [r.name for r in rules if r.parse().is_degenerate]
+    allow_slow = db.get_mquery_config_key("query_allow_slow") == "true"
```
Probably nitpicky, but should configuration really be in Redis? From my understanding Redis has a couple of responsibilities in this project: it's used as some sort of state management to keep the state of plugins, but also to store configuration, and there may be other uses I don't remember of the top of my head (like as a cache mechanism, possibly?)

Seems a bit cluttered. Maybe all config should be moved to a config file with a (simple) config parser implemented? Maybe something like in Karton. I quite like their [approach](https://github.com/CERT-Polska/karton/blob/master/karton/core/config.py#L152-L162) to loading booleans and such. Makes stuff like line 345 be as elegant as `config.get("query_allow_slow")` which would equate to 0 or 1

msm-code replied:
I don't think that's nitpicky! Quite the opposite, I think that's way too large issue to fit into a pull request comment :).

> but should configuration really be in Redis

Definitely not. Nothing should be in Redis except job scheduling internals (and maybe cache) (see also #74). Sadly we have to live with Redis for now, because I don't want to introduce another database at this stage of the project.

> Maybe all config should be moved to a config file with a (simple) config parser implemented? (...) Makes stuff like line 345 be as elegant as (...)

There are actually three issues hidden in this simple question:

**1. Refactor plugin configuration handling to remove the boilerplate**

This is already an issue here: #257. This is especially pressing for mquery's config fields. 

**2. Refactor mquery's config handling using typedconfig or configparser**

You may not think of it like this, but mquery already has a config file - [`config.py`](https://github.com/CERT-Polska/mquery/blob/66fba1b51490a40ca6c2978f6eaf99f350013b9d/src/config.docker.py). The intention of this file is to be modified with the deployment-specific config (this is why it's not tracked in git, for example). Things got more complicated because of docker - to make it more idiomatic for use in containers, we have config.docker.py that just reads all config fields from the environment (this is probably how you configure mquery).

I'm not fond of this pattern, but the current config is extremely simple (just backend URL, redis URL, forgotten "job expiration minutes" field #318 and plugin configuration).
I have it in the back of my mind to migrate someday to typedconfig or configparser, but (right now) it's a low priority refactoring and I'm weary of breaking backward compatibility.

**3. Move mquery core config fields from `plugin:Mquery` in the database to the config file**

(note: You say "all config should be moved", but I assume you mean all "core" config from `plugin:Mquery`. I don't think it's feasible to move plugin config to config file.)

After reading this I was initially dismissive of the idea, thinking it's unfeasible. After thinking a bit though, looks like [all current config fields](https://github.com/CERT-Polska/mquery/blob/66fba1b51490a40ca6c2978f6eaf99f350013b9d/src/db.py#L248) are viable candidates for a config file. This may be a bit harder for config fields that will influence worker behaviour (like #297 will), but still doable.

What I like about this idea:

* Less reinventing the whell, and fewer lines of code. Config refactor (#1) solved for free.
* Better separation of concerns - redis for plugins configuration, config file for mquery's core and web config.

What I don't like about this idea:

* This will break backward compatibility and we already have a release with redis-based mquery config. I think I can stomach that since I think number of orgs that integrated mquery auth with their orgs is very limited right now :).
* Impossible to configure mquery at runtime. Not necessarily a problem, current configuration keys should not be modified lightly and probably would benefit from being less interactive.
* At this point mquery's config gets a bit complicated, so we should combined this with (#2). I definitely don't want to have three configuration sources (redis, config.py, mquery.ini), therein lies madness. This means more backward-incompatible changes.

So wrapping up, I think you're onto something, but it needs a bit more design and it's definitely out of scope for this PR 🙃. Would you mind creating a separate issue for that?

ITAYC0HEN commented:
LGTM overall. Didn't check, but I hope that this feature works smartly with API requests as well, in some intuitive way.
Or is it?

msm-code commented:
> Didn't check, but I hope that this feature works smartly with API requests as well, in some intuitive way. Or is it?

Don't know about smartly, but it works:

* The `/api/query/` API has one new optional parameter - `force_slow_queries` (default is false). It works as expected - without it slow queries will be rejected. With the parameter set to true slow queries will be accepted (if mquery's configuration allows it).
* One caveat is that mquery error responses are not particularly API-friendly - just a JSON with an error message (no error code or something similar). So the API client doesn't know (without parsing the error message) what is the error cause (slow yara rule, syntax error, real internal server error, etc).
